### PR TITLE
Add config syncer for ASR with ACI

### PR DESCRIPTION
**Pull request by tbachman**
This adds the config syncer needed when the ASR
is used/integrated with ACI.

This closes issue #147

Signed-off-by: Thomas Bachman tbachman@yahoo.com

### CODE REVIEW

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+from networking_cisco.plugins.cisco.cfg_agent.device_drivers.asr1k import (
+    asr1k_cfg_syncer as syncer)
+from networking_cisco.plugins.cisco.common.htparser import HTParser
+
+
+LOG = logging.getLogger(__name__)
+
+
+#IP_REGEX = ^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: remove this line?

Comment (tbachman) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: Will fix.

File: networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+        self.config_syncer = aci_asr1k_cfg_syncer.ConfigSyncer(router_db_info,
+                                                      self.driver,
+                                                      self.hosting_device_info)
+        self.config_syncer.get_running_config = mock.Mock(
+            return_value=self._read_asr_running_cfg(
+                               'asr_basic_running_cfg_no_multi_region.json'))
+
+        invalid_cfg = self.config_syncer.delete_invalid_cfg()
+        self.assertEqual(8, len(invalid_cfg))
```
Comment (kentwu1) [inline networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py:127]: can we assert against the real list instead of just the length of it. This way u can also remove the chunk of comments above as it might go out of date easily....Same thing for other UTs.

File: networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        cfg.CONF.set_override('enable_multi_region', True, 'multi_region')
+        cfg.CONF.set_override('region_id', '0000002', 'multi_region')
+        cfg.CONF.set_override('other_region_ids', ['0000001'], 'multi_region')
+        router_db_info = []
+        self.config_syncer = aci_asr1k_cfg_syncer.ConfigSyncer(router_db_info,
+                                                      self.driver,
+                                                      self.hosting_device_info)
+        self.config_syncer.get_running_config = \
+            mock.Mock(return_value=self._read_asr_running_cfg(
+                                            'asr_basic_running_cfg.json'))
```
Comment (kentwu1) [inline networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py]: this asr_basic_running_cfg.json file doesn't contain config items tagged with region 0000001 so this UT doesn't really test the functionality that it will keep other region's config items....

File: networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+
+class AciASR1kCfgSyncer(test_sync.ASR1kCfgSyncer):
+
+    def _read_neutron_db_data(self):
+        """
+        helper function for reading the dummy neutron router db
+        """
+        with open(base.ROOTDIR +
+                  '/unit/cisco/etc/cfg_syncer/neutron_router_db_aci.json',
```
Comment (kentwu1) [inline networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py]: do we really need to add this file into this patch then load it up here? In the UTs below you always make the DB empty so it doesn't seem like we need this new file at all....

File: networking_cisco/tests/unit/cisco/etc/cfg_syncer/neutron_router_db_aci.json [hunk 5]
```diff
@@ -0,0 +1 @@
...aa1"}], "probe_connectivity": false, "priority": 100, "state": "ACTIVE", "redundancy_level": 2, "type": "HSRP"}, "id": "213fd355-78c3-44cc-bd6e-492b9ade2602", "share_host": true, "status": "ACTIVE", "routerhost:hosting_device": "00000000-0000-0000-0000-000000000003", "routes": [], "router_type": {"cfg_agent_service_helper": "neutron.plugins.cisco.cfg_agent.service_helpers.routing_svc_helper.RoutingServiceHelper", "cfg_agent_driver": "neutron.plugins.cisco.cfg_agent.device_drivers.asr1k.aci_asr1k_routing_driver.AciASR1kRoutingDriver", "name": "ASR1k_router", "id": "00000000-0000-0000-0000-000000000003"}, "routertype-aware-scheduler:auto_schedule": false, "cisco_ha:enabled": true, "hosting_device": {"status": "ACTIVE", "booting_time": 360, "host_category": "Hardware", "credentials": {"password": "cisco123", "user_name": "admin"}, "protocol_port": 22, "id": "00000000-0000-0000-0000-000000000003", "management_ip_address": "172.20.231.19", "service_types": "router:FW:VPN", "name": "ASR1k template", "admin_state_up": true, "created_at": "2015-11-16 22:54:18", "timeout": null, "template_id": "00000000-0000-0000-0000-000000000003"}, "routertype:id": "00000000-0000-0000-0000-000000000003"}]
```
Comment (kentwu1) [inline networking_cisco/tests/unit/cisco/etc/cfg_syncer/neutron_router_db_aci.json]: pre the comment earlier, is this file really needed?

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+                                           segment_nat_dict,
+                                           parsed_cfg)
+
+        if (cfg.CONF.multi_region.enable_multi_region):
+            set_route_regex = SET_ROUTE_MULTI_REGION_REGEX
+        else:
+            set_route_regex = SET_ROUTE_REGEX
+
+        invalid_cfg += self.clean_routes(conn,
+                                       router_id_dict,
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: the ident looks a bit off here....is pep8 still passing here?

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                                       intf_segment_dict,
+                                       segment_nat_dict,
+                                       parsed_cfg,
+                                       set_route_regex)
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: minor: add a empty line here as it seems that's convention inside this function?

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                     conn,
+                     router_id_dict,
+                     intf_segment_dict,
+                     segment_nat_dict,
+                     parsed_cfg,
+                     route_regex):
+        delete_route_list = []
+        routes = parsed_cfg.find_objects(route_regex)
+        for route in routes:
+            LOG.info(_LI("\ndefault route: %s"), (route))
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: why do we need bracket around 'route'?

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+        LOG.info(_LI("*************************"))
+        LOG.info(_LI("neutron router db records"))
+        for router_id, router in six.iteritems(router_id_dict):
+            LOG.info(_LI("ROUTER_ID: %s"), router_id)
+
+        LOG.info(_LI("\n"))
+
+        for segment_id, intf_list in six.iteritems(intf_segment_dict):
+            LOG.info(_LI("SEGMENT_ID: %s"), (segment_id))
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: why do we need bracket around 'segment_id'?

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+                     router_id_dict,
+                     intf_segment_dict,
+                     segment_nat_dict,
+                     parsed_cfg,
+                     route_regex):
+        delete_route_list = []
+        routes = parsed_cfg.find_objects(route_regex)
+        for route in routes:
+            LOG.info(_LI("\ndefault route: %s"), (route))
+            match_obj = re.match(route_regex, route.text)
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: why we have to apply the regex again here? earlier at line#133 we have done this: routes = parsed_cfg.find_objects(route_regex)
to get the routes....

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+            is_multi_region_enabled = cfg.CONF.multi_region.enable_multi_region
+            if (is_multi_region_enabled):
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: nit: don't need bracket here.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+                         {'router_id': router_id,
+                          'region_id': region_id,
+                          'segment_id': segment_id,
+                          'next_hop': next_hop})
+                my_region_id = cfg.CONF.multi_region.region_id
+                other_region_ids = cfg.CONF.multi_region.other_region_ids
+                if region_id != my_region_id:
+                    if region_id not in other_region_ids:
+                        delete_route_list.append(route.text)
+                    else:
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: don't need this else block as it will automatically do that.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+                             " %(segment_id)s, next_hop: %(next_hop)s") %
+                         {'router_id': router_id,
+                          'region_id': region_id,
+                          'segment_id': segment_id,
+                          'next_hop': next_hop})
+                my_region_id = cfg.CONF.multi_region.region_id
+                other_region_ids = cfg.CONF.multi_region.other_region_ids
+                if region_id != my_region_id:
+                    if region_id not in other_region_ids:
+                        delete_route_list.append(route.text)
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: this should also do "continue" to try the next route

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+                    else:
+                        # skip because some other deployment owns
+                        # this configuration
+                        continue
+            else:
+                router_id, segment_id, next_hop = (
+                    match_obj.group(1, 2, 3))
+
+            LOG.info(_LI("    router_id: %(router_id)s, segment_id:"
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: this log should be under the else block above (line#157) otherwise the same info will print it twice for multi-region case as there is also a similar log entry at line#141

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+
+            # Check that VRF exists in openstack DB info
+            if router_id not in router_id_dict:
+                LOG.info(_LI("router not found for route, deleting"))
+                delete_route_list.append(route.text)
+                continue
+
+            # Check that router has external network and segment_id matches
+            router = router_id_dict[router_id]
+            if "gw_port" not in router:
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: since you also call gw_port = router['gw_port'] at line#181, I think its more efficient to do things like this here:
gw_port = router.get('gw_port')
if gw_port is None:
   XXXXX
   continue

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+                LOG.info(_LI("router has no gw_port, route is invalid,"
+                             " deleting"))
+                delete_route_list.append(route.text)
+                continue
+
+            gw_port = router['gw_port']
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: we can then remove this line per previous comment

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+class ConfigSyncer(syncer.ConfigSyncer):
+
+    def __init__(self, router_db_info, driver,
+                 hosting_device_info, test_mode=False):
+        super(ConfigSyncer, self).__init__(router_db_info,
+                                           driver,
+                                           hosting_device_info,
+                                           test_mode=test_mode)
+
+    def delete_invalid_cfg(self, conn=None):
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: this function looks very similar to the one in the parent class. It seems to me that we just need to pass a different set of multi region regex to the clean_default_route()...We should be able to refactor the parents code a bit to achieve this instead of introducing all these duplicate codes.....

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+        invalid_cfg += self.clean_interfaces(conn,
+                                             intf_segment_dict,
+                                             segment_nat_dict,
+                                             parsed_cfg)
+
+        invalid_cfg += self.clean_vrfs(conn, router_id_dict, parsed_cfg)
+        LOG.debug("invalid_cfg = %s " % (pp.pformat(invalid_cfg)))
+        return invalid_cfg
+
+    def clean_routes(self,
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: this function is almost the same as clean_default_route() of the parent class....Can we just re-use the one in the parent's class then?

File: networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py (hunk 2 above)
Comment (tbachman) [inline networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py:127]: Will fix.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py (hunk 6 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: Strangely it passed, but will fix.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py (hunk 17 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: Yes -- will refactor.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py (hunk 18 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_cfg_syncer.py]: Will refactor.

File: networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py (hunk 4 above)
Comment (tbachman) [inline networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py]: We need a DB file. The reason is that some of the tests from the parent class use the DB.

File: networking_cisco/tests/unit/cisco/etc/cfg_syncer/neutron_router_db_aci.json (hunk 5 above)
Comment (tbachman) [inline networking_cisco/tests/unit/cisco/etc/cfg_syncer/neutron_router_db_aci.json]: I will investigate whether we can use the existing DB file -- the only difference between this DB file and the existing one is that this passes the correct configuration agent driver (AciASR1k, instead of ASR1k). However, I don't think this ever gets used, and if that's the case, we'll just use the existing DB file.

File: networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py (hunk 3 above)
Comment (tbachman) [inline networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py]: Will investigate and fix if that's the case.

Comment (tbachman) [inline networking_cisco/tests/unit/cisco/cfg_agent/test_aci_asr1k_cfg_syncer.py]: Changing to asr_running_cfg.json, which has those elements, and fixing the test to account for the additional config items.

File: networking_cisco/tests/unit/cisco/etc/cfg_syncer/neutron_router_db_aci.json (hunk 5 above)
Comment (tbachman) [inline networking_cisco/tests/unit/cisco/etc/cfg_syncer/neutron_router_db_aci.json]: Changed to existing DB file.

**Comment by tbachman** [comment]
Addressed reviewer comments with new pull request.
